File: AIReader/AIReader/Services/Sync/WebDAV.swift
import Foundation

/// A file or folder a WebDAV server lists.
struct WebDAVEntry: Equatable, Sendable, Identifiable {
    let url: URL
    let isFolder: Bool

    var id: URL { url }
    var name: String { url.lastPathComponent }
}

/// What the app needs of a WebDAV server: fetch one file, store one file,
/// and list a folder. Any server that speaks HTTP GET, PUT and PROPFIND with
/// a password will do.
enum WebDAV {
    enum DAVError: LocalizedError {
        case unauthorized
        case http(status: Int)

        var errorDescription: String? {
            switch self {
            case .unauthorized: "The server refused the user name or password."
            case let .http(status): "The server answered \(status)."
            }
        }
    }

    /// The file's bytes, or nil when there is no such file yet.
    static func download(_ url: URL, settings: SyncSettings) async throws -> Data? {
        let (data, status) = try await send("GET", url, body: nil, settings: settings)
        if status == 404 { return nil }
        try check(status)
        return data
    }

    /// Stores the file, making its folder first if the server says there is
    /// none.
    static func upload(
        _ data: Data,
        to url: URL,
        type: String = "application/json",
        settings: SyncSettings
    ) async throws {
        let (_, status) = try await send("PUT", url, body: data, type: type, settings: settings)
        if status == 409 || status == 404 {
            try await makeFolder(url.deletingLastPathComponent(), settings: settings)
            let (_, again) = try await send("PUT", url, body: data, type: type, settings: settings)
            try check(again)
            return
        }
        try check(status)
    }

    /// The folder's own entries, folders first.
    static func list(_ folder: URL, settings: SyncSettings) async throws -> [WebDAVEntry] {
        let (data, status) = try await send("PROPFIND", folder, body: nil, depth: "1", settings: settings)
        try check(status)
        var entries: [WebDAVEntry] = []
        var href = ""
        var isFolder = false
        func close() {
            if let url = URL(string: href, relativeTo: folder)?.absoluteURL,
               url.standardized.path != folder.standardized.path {
                entries.append(WebDAVEntry(url: url, isFolder: isFolder))
            }
            href = ""
            isFolder = false
        }
        XMLScanner.scan(data) { name, _ in
            if name == "response", !href.isEmpty { close() }
            if name == "collection" { isFolder = true }
        } onText: { name, text in
            if name == "href" { href = text.trimmingCharacters(in: .whitespacesAndNewlines) }
        }
        if !href.isEmpty { close() }
        return entries.sorted { ($0.isFolder ? 0 : 1, $0.name) < ($1.isFolder ? 0 : 1, $1.name) }
    }

    /// Makes the folder, and the folders above it that are missing too: a
    /// server makes only one level at a time, answering 409 when the one
    /// above is not there.
    private static func makeFolder(_ url: URL, settings: SyncSettings, depth: Int = 0) async throws {
        let (_, status) = try await send("MKCOL", url, body: nil, settings: settings)
        guard status == 409, depth < 8 else { return }
        try await makeFolder(url.deletingLastPathComponent(), settings: settings, depth: depth + 1)
        _ = try await send("MKCOL", url, body: nil, settings: settings)
    }

    private static func send(
        _ method: String,
        _ url: URL,
        body: Data?,
        type: String = "application/json",
        depth: String? = nil,
        settings: SyncSettings
    ) async throws -> (Data, Int) {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = body
        request.timeoutInterval = 30
        if let depth { request.setValue(depth, forHTTPHeaderField: "Depth") }
        if !settings.username.isEmpty || !settings.password.isEmpty {
            let credentials = Data("\(settings.username):\(settings.password)".utf8).base64EncodedString()
            request.setValue("Basic \(credentials)", forHTTPHeaderField: "Authorization")
        }
        if body != nil {
            request.setValue(type, forHTTPHeaderField: "Content-Type")
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        return (data, (response as? HTTPURLResponse)?.statusCode ?? 0)
    }

    private static func check(_ status: Int) throws {
        if status == 401 || status == 403 { throw DAVError.unauthorized }
        guard (200...299).contains(status) else { throw DAVError.http(status: status) }
    }
}

File: BrowserExtension/lib/webdav.js
// The WebDAV folder AIReader syncs through. Books go into its `Books`
// folder, where both apps fetch them on their next sync.

export const booksFolder = "Books";

/** The folder URL, with a trailing slash. */
function folderURL(settings, ...names) {
  const base = settings.url.trim().replace(/\/+$/, "");
  return [base, ...names.map(encodeURIComponent)].join("/") + "/";
}

async function send(method, url, settings, { body, headers = {} } = {}) {
  const credentials = `${settings.username}:${settings.password}`;
  const bytes = new TextEncoder().encode(credentials);
  const authorization = settings.username || settings.password
    ? { Authorization: `Basic ${btoa(String.fromCharCode(...bytes))}` }
    : {};
  const response = await fetch(url, { method, body, headers: { ...authorization, ...headers }, credentials: "omit" });
  if (response.status === 401 || response.status === 403) {
    throw new Error("The server refused the user name or password.");
  }
  return response;
}

function check(response) {
  if (!response.ok) throw new Error(`The server answered ${response.status}.`);
}

/** The names of the files in the books folder; none if it is not there yet. */
export async function listBooks(settings) {
  const response = await send("PROPFIND", folderURL(settings, booksFolder), settings, { headers: { Depth: "1" } });
  if (response.status === 404) return [];
  check(response);
  const xml = new DOMParser().parseFromString(await response.text(), "application/xml");
  const names = [];
  for (const entry of xml.getElementsByTagNameNS("DAV:", "response")) {
    if (entry.getElementsByTagNameNS("DAV:", "collection").length > 0) continue;
    const href = entry.getElementsByTagNameNS("DAV:", "href")[0]?.textContent.trim() ?? "";
    const name = decodeURIComponent(href.replace(/\/+$/, "").split("/").pop());
    if (name) names.push(name);
  }
  return names;
}

/**
 * Makes the folder, and the folders above it that are missing too: a
 * server makes only one level at a time, answering 409 when the one above
 * is not there.
 */
async function makeFolder(url, settings, depth = 0) {
  const response = await send("MKCOL", url, settings);
  if (response.status !== 409 || depth > 8) return;
  await makeFolder(new URL("..", url).href, settings, depth + 1);
  await send("MKCOL", url, settings);
}

/** Stores a book in the books folder under `name`. */
export async function putBook(name, blob, settings) {
  const folder = folderURL(settings, booksFolder);
  const url = folder + encodeURIComponent(name);
  const options = { body: blob, headers: { "Content-Type": "application/epub+zip" } };
  let response = await send("PUT", url, settings, options);
  if (response.status === 409 || response.status === 404) {
    await makeFolder(folder, settings);
    response = await send("PUT", url, settings, options);
  }
  check(response);
}

/** Whether the folder is there; throws when the server cannot be reached or refuses. */
export async function checkFolder(settings) {
  const response = await send("PROPFIND", folderURL(settings), settings, { headers: { Depth: "0" } });
  if (response.status === 404) return false;
  check(response);
  return true;
}
